Thick Arrays

vkbo opened the pull request:
There were 30 arrays in `mod_common_main` that were only used in `synuthck`, where they were initialised inside particle loops before every time they were used. That is, at no point were the values accessed elsewhere than within the loop and on the same particle. In other words, they do not need to be arrays.

In fact, this is how it is done in the routine `envars`. In `envarsv` the arrays are allocated locally, but also here the values are never accessed outside of loops, so also in this routine the arrays are redundant.

Each commit in the PR addresses one array, or a set of related arrays. It may be easier to read the diff on a commit by commit basis.

kyrsjo commented on source/sixve.f90:
```diff
@@ -211,397 +211,365 @@ subroutine envarsv(dpsv,oidpsv,rvv,ekv)
+    si,siq,sm1,sm12,sm2,sm23,sm3,wf,wfa,wfhi,fokm
 
-!-----------------------------------------------------------------------
   save
```
Do we need to `save` this, thus keeping the arrays always-in-memory in case the function gets called again?

Also, isn't `npart` dynamic now? How does that work when you declare a static e.g. `dpd(npart)`?

vkbo replied:
I assumed the save was needed for the remaining arrays, which is only a few now, so it's not much of an issue. Although I could test if removing save blows up.

kyrsjo commented on source/sixve.f90:
```diff
@@ -211,397 +211,365 @@ subroutine envarsv(dpsv,oidpsv,rvv,ekv)
+    if(abs(el(l)) <= pieni) goto 160
+
+    kz1 = kz(l)+1
+    if(kz1 == 1) goto 20
```
Select case? Elseif?

vkbo replied:
Can't because the goto labels are in use also aside from selecting on kz

kyrsjo commented on source/sixve.f90:
```diff
@@ -211,397 +211,365 @@ subroutine envarsv(dpsv,oidpsv,rvv,ekv)
+      as(4,ih1,j,l) = as4+(two*as6)*fok1
+      as(5,ih1,j,l) = (as6*fok1**2-(rvv(j)*sm12)/(c4e3*rho**2))+fok1*as4
+      as(6,ih1,j,l) = as6
 !--VERTIKAL
```
Degermanizing required :)

vkbo replied:
I de-germanised and norwegianised it. True story.

kyrsjo commented on source/sixve.f90:
```diff
@@ -211,397 +211,365 @@ subroutine envarsv(dpsv,oidpsv,rvv,ekv)
     goto 160
 !-----------------------------------------------------------------------
 !  RECTANGULAR MAGNET
 !  HORIZONTAL
```
Vertical or horizontal? Comments are strange here...

vkbo replied:
No idea.

kyrsjo commented on source/sixve.f90:
```diff
@@ -211,397 +211,365 @@ subroutine envarsv(dpsv,oidpsv,rvv,ekv)
     goto 160
 !-----------------------------------------------------------------------
 !  SEKTORMAGNET
 !  HORIZONTAL
```
There are also vertical ones here...

kyrsjo commented on source/sixve.f90:
```diff
@@ -211,397 +211,365 @@ subroutine envarsv(dpsv,oidpsv,rvv,ekv)
+      as(4,ih1,j,l) = ((-one*rvv(j))*sm23)/c2e3
+      as(5,ih1,j,l) = ((-one*rvv(j))*sm12)/(c4e3*rho**2)
+      as(6,ih1,j,l) = ((-one*rvv(j))*(el(l)+al(1,ih1,j,l)*al(2,ih1,j,l)))/c4e3
 !--VERTIKAL
```
More german...

kyrsjo commented on source/sixve.f90:
```diff
@@ -211,397 +211,365 @@ subroutine envarsv(dpsv,oidpsv,rvv,ekv)
     goto 160
 !-----------------------------------------------------------------------
 !  QUADRUPOLE
 !  FOCUSSING
```
Splelling

kyrsjo replied:
Also, defocusing is further down in the same block...

kyrsjo commented on source/track_thick.f90:
```diff
@@ -1886,317 +1886,324 @@ subroutine synuthck
       goto 20
 !-----------------------------------------------------------------------
 !  RECTANGULAR MAGNET
 !  HORIZONTAL
```
See comments about comments in the previous file. I won't repeat this further...

kyrsjo commented on source/track_thick.f90:
```diff
@@ -1886,317 +1886,324 @@ subroutine synuthck
       goto 160
     else
 !Eric
 ! Is really an error but old code went to 160
```
If it is an error, then print a warning at least? Or the comment is wrong and kz can be not in the range of the else-ifs...

vkbo replied:
Feel free to investigate whether Eric is correct, and add this. The thick tracking code is not well maintained, and not all that well covered in tests. Although most of this routine is covered.

vkbo commented:
Ok, so, after your comments @kyrsjo , I removed the temp arrays in `envarsv` and loaded the global ones instead. See my last few commits that do them one by one. I also removed the save flag, which is now trivial anyway since there are no arrays left.

The dpd and dpsq arrays are also now the global ones, and should be fine. The fokqv one is computed in the exact same way as in `thick_arrays`, which is `+/-ekv`.

kyrsjo commented:
I'll heard you like PRs, so I'm making a PR to your PR!

vkbo commented:
> I'll heard you like PRs, so I'm making a PR to your PR!

Careful, or I'll PR your PR to my PR.

kyrsjo commented:
Wait, I need to PR my PR so you can PR your PR with my PR!

kyrsjo commented:
Plz mrg. https://github.com/vkbo/SixTrack/pull/2
